jupyter-extension/jupyterlab_pachyderm/pfs_manager.py

```diff
@@ -214,16 +219,325 @@ def get(self, path, content=True, type=None, format=None) -> dict:
+        # self.next_datum()
+
+    # TODO: don't ignore name in the input spec
+    # right now, when a repo is mounted as part of mounting datum(s), the
+    # toplevel directory for the repo is project_repo_branch. we should
+    # update this to respect the name in the pfsinput objects being supplied
+    # to the mount_datums api. however, this is not that simple as the name
+    # can't simply be looked up in the datuminfo. we will need to have some
+    # mapping to track which files within a datum belong to which name.
+    def mount_datums(self, input_dict: dict):
+        input = pps.Input().from_dict(input_dict["input"])
+        datum = self._client.pps.list_datum(input=input)
+        self._datum_list = [d for d in datum]
+        self._datum_index = 0
+        self._mount_time = datetime.datetime.now()
+        if len(self._datum_list) == 0:
+            raise ValueError("input produced no datums to mount")
+        self._update_mount()
+
+    def next_datum(self):
+        self._datum_index = (self._datum_index + 1) % len(self._datum_list)
+        self._update_mount()
+
+    def prev_datum(self):
+        if self._datum_index == 0:
+            self._datum_index = len(self._datum_list)
+        self._datum_index = self._datum_index - 1
+        self._update_mount()
+
+    def _get_fileinfo_path(self, fileinfo: pfs.FileInfo) -> Path:
+        project = fileinfo.file.commit.repo.project.name
+        branch = fileinfo.file.commit.branch.name
+        repo = fileinfo.file.commit.repo.name
+        toplevel = f"{project}_{repo}_{branch}"
+        return Path(self._FILEINFO_DIR, toplevel, fileinfo.file.path.strip("/"))
+
+    def _update_mount(self):
+        os.system(f"rm -rf {self._FILEINFO_DIR}/*")
+        self._dirs.clear()
+        for fileinfo in self._datum_list[self._datum_index].data:
+            path = self._get_fileinfo_path(fileinfo=fileinfo)
+
+            if path.exists():
+                if path.is_dir():
+                    # the dir being mounted is the parent of a file that is already mounted
+                    os.system(f"rm -rf {path}")
+                else:
+                    return
+
+            os.makedirs(str(path.parent), exist_ok=True)
+            with open(path, "w") as f:
+                f.write(fileinfo.to_json())
+            if fileinfo.file_type == pfs.FileType.DIR:
+                self._dirs.add(str(path))
+
+    def _get_file_from_path(self, path: str) -> pfs.File:
+        path = path.strip("/")
+        if not path:
+            return None
+
+        parts = Path(path).parts
+        project, repo, branch = parts[0].split("_")
+        if len(parts) == 1:
+            pach_path = "/"
+        else:
+            pach_path = str(Path(*parts[1:]))
+        file_uri = f"{project}/{repo}@{branch}:{pach_path}"
+        return pfs.File.from_uri(file_uri)
+
+    def is_hidden(self, path):
+        return False
+
+    def file_exists(self, path) -> bool:
```

Comment:
What is the criteria for a file or directory to exist? That it's part of a datum? It exists on disk?

Reply:
in this context, a file or directory "exists" when it is part of the datum being mounted. the comment here explains the conditions under which a file would exist, and the comment in the next method explains the conditions under which a directory would exist.